### ISSUE TITLE
App resource deployment always fails when SSH is disabled on space level

### ISSUE THREAD

**Issue by oysteing** [cloudfoundry-community/terraform-provider-cloudfoundry#384]
Given the following resource:
```
data "cloudfoundry_space" "target" {
  name = "space-with-ssh-disabled"
  org_name = "sample-org"
}

resource "cloudfoundry_app" "ssh-test" {
  name             = "ssh-test"
  space            = data.cloudfoundry_space.target.id
  path             = "sample.jar"
  buildpack        = "java_buildpack_offline"
  enable_ssh       = false
}
```

The deployment always fails with a 400 Bad request from Cloud Foundry:
```
2022-03-28T08:02:34.941+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/28 08:02:34 {
  "code": 10004,
  "description": "The request is invalid",
  "error_code": "CF-InvalidRequest"
}: timestamp=2022-03-28T08:02:34.940+0200
2022-03-28T08:02:34.942+0200 [WARN]  unexpected data: registry.terraform.io/cloudfoundry-community/cloudfoundry:stderr="{"@caller":"github.com/hashicorp/terraform-plugin-sdk/v2@v2.12.0/internal/logging/helper_schema.go:21","@level":"trace","@message":"Called downstream","@module":"sdk.helper_schema","@timestamp":"2022-03-28T08:02:34.940708+02:00"}"
2022-03-28T08:02:34.964+0200 [ERROR] vertex "cloudfoundry_app.ssh-test" error: The request is invalid
╷
│ Error: The request is invalid
│ 
│   with cloudfoundry_app.ssh-test,
│   on app.tf line 6, in resource "cloudfoundry_app" "ssh-test":
│    6: resource "cloudfoundry_app" "ssh-test" {
│ 
╵
2022-03-28T08:02:34.990+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-03-28T08:02:34.995+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudfoundry-community/cloudfoundry/0.15.2/darwin_amd64/terraform-provider-cloudfoundry_v0.15.2 pid=66277
2022-03-28T08:02:34.995+0200 [DEBUG] provider: plugin exited

```

See also https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/issues/273

From my point of view, the if block https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/blob/master/cloudfoundry/structures_app.go#L17 can be completely removed to fix the problem.

**Comment by loafoe** [comment]
@oysteing based on the output it looks like the provider is crashing. Are you able to run the provider with the `--debug` flag? It will output something like this:

```bash
TF_REATTACH_PROVIDERS='{"registry.terraform.io/cloudfoundry-community/cloudfoundry":{"Protocol":"grpc","ProtocolVersion":5,"Pid":15714,"Test":true,"Addr":{"Network":"unix","String":"/var/folders/..."}}}'
```

By setting this variable in the environment where you run `terraform apply` (locally) it will use the binary under debug. If it indeed crashes you should see a stack trace which will help pinpoint the issue.

**Comment by oysteing** [comment]
I can't really see that the provider is crashing, but the cloud controller replies with 400 Bad request to the POST /v2/apps request. I don't find any --debug flag on the provider, but I enabled cloudfoundry-cli logging with `CF_LOG=debug terraform apply`. Here the log output:
```
cloudfoundry_app.ssh-test: Creating...
2022-03-31T09:17:57.313+0200 [INFO]  Starting apply for cloudfoundry_app.ssh-test
2022-03-31T09:17:57.313+0200 [DEBUG] cloudfoundry_app.ssh-test: applying the planned Create change
2022-03-31T09:17:57.315+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 [DEBUG] setting computed for "ports" from ComputedKeys: timestamp=2022-03-31T09:17:57.315+0200
2022-03-31T09:17:57.315+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 [DEBUG] setting computed for "environment" from ComputedKeys: timestamp=2022-03-31T09:17:57.315+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 [INFO] Use deploy strategy standard: timestamp=2022-03-31T09:17:57.315+0200
2022-03-31T09:17:57.316+0200 [WARN]  unexpected data: registry.terraform.io/cloudfoundry-community/cloudfoundry:stderr="{"@caller":"github.com/hashicorp/terraform-plugin-sdk/v2@v2.12.0/internal/logging/helper_schema.go:21","@level":"trace","@message":"Calling downstream","@module":"sdk.helper_schema","@timestamp":"2022-03-31T09:17:57.315542+02:00"}"
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 REQUEST: [2022-03-31T09:17:57+02:00]: timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 POST /v2/apps HTTP/1.1: timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 host: [masked]: timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Accept: application/json: timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Content-Type: application/json: timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 User-Agent: terraform-provider/0.0.0-unknown-version (go1.18; amd64 darwin): timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.316+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 {
  "buildpack": "java_buildpack_offline",
  "enable_ssh": true,
  "health_check_http_endpoint": "",
  "health_check_type": "port",
  "instances": 1,
  "memory": 1,
  "name": "ssh-test",
  "ports": [
    8080
  ],
  "space_guid": "f155ed64-6428-45b3-8647-3e29e0d9808f",
  "state": "STOPPED"
}: timestamp=2022-03-31T09:17:57.316+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 RESPONSE: [2022-03-31T09:17:57+02:00]: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 HTTP/1.1 400 Bad Request: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Connection: keep-alive: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Content-Length: 100: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Content-Type: application/json;charset=utf-8: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Date: Thu, 31 Mar 2022 07:17:57 GMT: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 Server: nginx: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 X-B3-Spanid: 3e677fc6798537bd6784f35a77792f0f: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 X-B3-Traceid: 3e677fc6798537bd6784f35a77792f0f: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 X-Content-Type-Options: nosniff: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 X-Vcap-Request-Id: 50b4424b-1b8c-4455-4ea3-a7f7a04b713e::bfee836f-642e-4a78-847d-94ce72130792: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.560+0200 [INFO]  provider.terraform-provider-cloudfoundry_v0.15.2: 2022/03/31 09:17:57 {
  "code": 10004,
  "description": "The request is invalid",
  "error_code": "CF-InvalidRequest"
}: timestamp=2022-03-31T09:17:57.560+0200
2022-03-31T09:17:57.562+0200 [WARN]  unexpected data: registry.terraform.io/cloudfoundry-community/cloudfoundry:stderr="{"@caller":"github.com/hashicorp/terraform-plugin-sdk/v2@v2.12.0/internal/logging/helper_schema.go:21","@level":"trace","@message":"Called downstream","@module":"sdk.helper_schema","@timestamp":"2022-03-31T09:17:57.560424+02:00"}"
2022-03-31T09:17:57.582+0200 [ERROR] vertex "cloudfoundry_app.ssh-test" error: The request is invalid
╷
│ Error: The request is invalid
│ 
│   with cloudfoundry_app.ssh-test,
│   on app.tf line 6, in resource "cloudfoundry_app" "ssh-test":
│    6: resource "cloudfoundry_app" "ssh-test" {
│ 
╵
2022-03-31T09:17:57.605+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-03-31T09:17:57.607+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudfoundry-community/cloudfoundry/0.15.2/darwin_amd64/terraform-provider-cloudfoundry_v0.15.2 pid=39366
2022-03-31T09:17:57.607+0200 [DEBUG] provider: plugin exited
```

To me it looks like the critical part is that `"enable_ssh": true` is sent although the resource is configured with `enable_ssh = false`.

**Comment by loafoe** [comment]
@oysteing oh sorry should have been more clear. The `--debug` flag is for the Cloud foundry provider binary. This binary should be somewhere in the folder:

`.terraform/providers/registry.terraform.io/cloudfoundry-community/cloudfoundry/0.15.2/darwin_arm64`

(path depends on your version and arch)

You can start this binary with the `--debug` flag and then use the `TF_REATTACH_PROVIDERS` value in the environment where you run `terraform apply`

**Comment by oysteing** [comment]
Ok. This time it worked. I ran the commands:
`.terraform/providers/registry.terraform.io/cloudfoundry-community/cloudfoundry/0.15.2/darwin_amd64/terraform-provider-cloudfoundry_v0.15.2 --debug 2> terraform-provider-cloudfoundry_v0.15.2_debug.log
TF_REATTACH_PROVIDERS='{"registry.terraform.io/cloudfoundry-community/cloudfoundry":{"Protocol":"grpc","ProtocolVersion":5,"Pid":44487,"Test":true,"Addr":{"Network":"unix","String":"/var/folders/ty/8d57kj253636l0r0d4c2x_240000gq/T/plugin2272888900"}}}' terraform apply`

The debug log file: [terraform-provider-cloudfoundry_v0.15.2_debug.log](https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/files/8388710/terraform-provider-cloudfoundry_v0.15.2_debug.log)

**Comment by loafoe** [comment]
Ok, nothing specific in the debug log. I looked at the `diff` you mentioned and I also don't really understand the logic behind it. I'll do some test runs with the block disabled to see if there are any ill effects

**Comment by loafoe** [comment]
@oysteing I pushed a possible fix for this issue. I've tested this briefly on one of our CF clusters, seems to work where it failed earlier with your example code. If you are able to test the branch out that would be great! (go build the branch and then run the binary with `--debug`)

**Comment by oysteing** [comment]
With this branch it works! Here's the debug log: [terraform-provider-cloudfoundry_bugfix_384_debug.log](https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/files/8393240/terraform-provider-cloudfoundry_bugfix_384_debug.log)
 
I can't say I understand all aspects of this piece of code, but I would be very happy to see this fix.

**Comment by loafoe** [comment]
As you mentioned, the original code forced `enable_ssh` for new resources which broke the API call for spaces which have this disabled 100%. The updated code honours the settings and also leaves out any value if it's not specified (accepting the CF server settings)

The change is available in `0.15.3`
